pkg/ocm/config.go

```diff
@@ -39,6 +42,46 @@ var URLAliases = map[string]string{
 	"crc":         "https://clusters-service.apps-crc.testing",
 }
 
+func ValidOCMUrlAliases() []string {
+	keys := make([]string, 0, len(URLAliases))
+	for k := range URLAliases {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+// URL Precedent (from highest priority to lowest priority):
+//  1. runtime `--env` cli arg (key found in `urlAliases`)
+//  2. runtime `--env` cli arg (non-empty string)
+//  3. config file `URL` value (non-empty string)
+//  4. sdk.DefaultURL
+//
+// Finally, it will try to url.ParseRequestURI the resolved URL to make sure it's a valid URL.
+func ResolveGatewayURL(optionalParsedCliFlagValue string, optionalParsedConfig *config.Config) (string, error) {
+	gatewayURL := sdk.DefaultURL
+	source := "default"
+	if optionalParsedCliFlagValue != "" {
+		gatewayURL = optionalParsedCliFlagValue
+		source = "flag"
+		if _, ok := URLAliases[optionalParsedCliFlagValue]; ok {
+			gatewayURL = URLAliases[optionalParsedCliFlagValue]
+		}
+	} else if optionalParsedConfig != nil && optionalParsedConfig.URL != "" {
+		// re-use the URL from the config file
+		gatewayURL = optionalParsedConfig.URL
+		source = "config"
+	}
+
+	url, err := url.ParseRequestURI(gatewayURL)
+	if err != nil {
+		return "", fmt.Errorf(
+			"%w\n\nURL Source: %s\nExpected an absolute URI/path (e.g. %s) or a case-sensitive alias, one of: [%s]",
+			err, source, sdk.DefaultURL, strings.Join(ValidOCMUrlAliases(), ", "))
+	}
+
+	return url.String(), nil
+}
+
 func GetEnv() (string, error) {
```

Comment:
how is this used in rosa? it probably needs to be made compatible with the `--rh-region` url override. I.e. what happens when `GetEnv()` is called after someone does `rosa login --rh-region=ap-southeast-1` and the URL in config lookes like "https://api.aws.ap-southeast-1.stage.openshift.com" im pretty sure it will return nil/err and whatever command is relying on it will break.

Reply:
in otherwords im saying this function needs to be modified so that cfg.URL=https://api.aws.ap-southeast-1.stage.openshift.com/ returns "stage", etc.

Reply:
ACK, we need to ensure there is an alias for every url

Reply:
Just to be clear, region discovery means we cant hard code aliases for every regional url. the point is for us to be able to add regions without requiring clients to update their cli version. However, we can map regions to known env aliases using naming conventions, e.g. https://api.aws.ap-southeast-1.stage.openshift.com is in the "stage" env. We actually already basically do this in the ocm-sdk when we are determining which global gateway url to use for region discovery: https://github.com/openshift-online/ocm-sdk-go/blob/main/rh_region.go#L48

Similar tactic should work in GetEnv()

Reply:
Added EnvAliases to ensure we are able to discover regionalized environments. However, haven't added production URL (`openshift.com`) yet, since we don't have shards there. Also, comparing and returning **production** env based on just the suffix `openshift.com` does not seem like a concrete check for now (since `openshift.com` is a suffix for **int**(`integration.openshift.com`) and **stage**(`stage.openshift.com`) environments as well). 
We do need to tackle this though. My suggestion is to wait till we define regionalized production URLs and have a unique signature there that we can use here. We might also consider defaulting a production env but that seems a bit risky.
Thoughts @macgregor?

Reply:
Can't we just push this into the `ocm-shards.json` as another field alongside `url` so we have a single source of truth on the environment `name`, `url` and `alias`?

Reply:
@tirthct we dont need to wait on anything, ocm-shards.json exists in prod, its just an empty json structure. ocm-sdk-go and ocm-cli both handle it fine:
```
> ocm list rh-regions                     
Discovery URL: https://api.openshift.com

RH Region    Gateway URL
```

Reply:
> Can't we just push this into the ocm-shards.json as another field alongside url so we have a single source of truth on the environment name, url and alias?

1. rosa is the only thing that cares about the environment name. 
2. to get ocm-shards.json you necessarily need to know which environment to ask for the discovery file:
  * https://api.integration.openshift.com/static/ocm-shards.json
  * https://api.stage.openshift.com/static/ocm-shards.json
  * https://api.openshift.com/static/ocm-shards.json
3. as ive mentioned a few times, we already have a shared way of determining what gateway/discovery file should be used from an arbitrary url: https://github.com/openshift-online/ocm-sdk-go/blob/main/rh_region.go#L48. I am not changing the ocm-shards.json schema right now to add the name of the environment it was fetched from since the client (rosa) already knows it by then. 

So the answer here is to use the shared DetermineRegionDiscoveryUrl from ocm-sdk to go from arbitrary ocm url (global, regional, int, stage, prod, whatever) to global discovery gateway url (i.e. the int, stage, prod ocm urls you know today). Display it or refer to that as "prod" or "api.openshift.com" makes no difference and is trivial to map client side, and what we will do here: https://github.com/openshift/rosa/pull/1655/files#r1516218065